### ISSUE TITLE
Ability to generate Java annotations with Class<...>[] field types

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon-compiler#2122]
Java annotations may have fields of type `Class<...>[]`, and it may be useful to define such annotations in Ceylon for interop with Java APIs such as `javax.validation`, that per <http://beanvalidation.org/1.0/spec/#constraintsdefinitionimplementation-constraintdefinition-properties> requires:

```java
	Class<?>[] groups() default {};
	Class<? extends Payload>[] payload() default {};
 ```

The closest approximation I’ve found in Ceylon is:

```ceylon
	shared ClassOrInterfaceDeclaration[] groups = []
```

but compiles to `String[]` (IIRC).

This was originally [discussed](https://groups.google.com/forum/#!msg/ceylon-dev/HTPUVzc83hc/MJuPvAiLLQwJ) on the dev mailing list.

**Comment by tombentley** [comment]
The problem here is that the Ceylon typechecker has strict rules about the types of attributes used in a Ceylon annotation class, and `java.lang.Class` isn't one of those types. Declaration references (including `ClassOrInterfaceDeclaration`) are encoded as `String`s. I suppose we could change that so that if the type of an attribute of an annotation class was precisely `ClassOrInterfaceDeclaration` we could use a `java.lang.Class<?>` instead. Doing that would mean changing the declaration reference type would become a binary incompatible change, which isn't ideal.

That wouldn't be enough for the `payload()` example above though, because there's no way to express the upper bound. To support that we would need to support meta models in annotation classes (which is https://github.com/ceylon/ceylon-spec/issues/847, but it's the compiler which is the blocker), allowing you to write:

    shared ClassOrInterface<out Payload>[] payload = [];

Then we could compile that to a `Class<? extends Payload>`.

So in principle this is a soluble problem. But I worry that the way we transform this stuff is getting pretty complicated, ad-hoc and therefore unintuitive.

**Comment by FroMage** [comment]
Don't we already have a special case to be able to produce a `Class<?>` in Java annotations?

**Comment by tombentley** [comment]
Yes, but this is about *declaring* an annotation type with a `j.l.Class` typed parameter, not using one. The typechecker doesn't know anything about `j.l.Class`, but it has to know about the valid annotation constructor types.

**Comment by FroMage** [comment]
Ah, OK. Well yes, then it requires support from the typechecker then.

**Comment by tombentley** [comment]
Well not really, assuming the solution I outlined above is satisfactory. 

I have the declaration (as opp. model) part of it working locally, but I've not pushed it because it would definitely break compatibility with 1.1.

**Comment by gavinking** [comment]
>  Declaration references (including `ClassOrInterfaceDeclaration`) are encoded as `String`s. I suppose we could change that so that if the type of an attribute of an annotation class was precisely `ClassOrInterfaceDeclaration` we could use a `java.lang.Class<?>` instead. 

I guess I would be in favor of that change.

**Comment by FroMage** [comment]
@tombentley since we broke compat with other annotation changes, you want to merge this?

**Comment by tombentley** [comment]
Yeah, I'll try and dig it out and rebase it later.

**Comment by FroMage** [comment]
Moving to 1.3.
